File: Day51/Part4/virus_containment.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <set>

using namespace std;

const int INF = 1e9;

struct Edge {
    int to;
    int capacity;
    int flow;
    int rev; // index of reverse edge
};

vector<vector<Edge>> adj;
vector<int> level;
vector<int> ptr;

void add_edge(int from, int to, int cap) {
    Edge a = {to, cap, 0, (int)adj[to].size()};
    Edge b = {from, cap, 0, (int)adj[from].size()}; // Undirected edge: cap in both directions?
    // Yes, for min-cut in undirected graph, we treat edge as two directed edges with capacity C.
    // Wait, if we push flow u->v, we use capacity.
    // If we push v->u, we use capacity.
    // This models the fact that the connection can carry information in either direction,
    // but cutting it removes BOTH.
    // Actually, standard reduction for undirected min-cut is replace (u,v) with u->v (cap) and v->u (cap).
    adj[from].push_back(a);
    adj[to].push_back(b);
}

// For Source/Sink edges which are directed (S->Node, Node->T)
void add_directed_edge(int from, int to, int cap) {
    Edge a = {to, cap, 0, (int)adj[to].size()};
    Edge b = {from, 0, 0, (int)adj[from].size()}; // Reverse edge has 0 capacity
    adj[from].push_back(a);
    adj[to].push_back(b);
}

int bfs(int s, int t) {
    fill(level.begin(), level.end(), -1);
    level[s] = 0;
    queue<int> q;
    q.push(s);
    while (!q.empty()) {
        int v = q.front();
        q.pop();
        for (auto& edge : adj[v]) {
            if (edge.capacity - edge.flow > 0 && level[edge.to] == -1) {
                level[edge.to] = level[v] + 1;
                q.push(edge.to);
            }
        }
    }
    return level[t] != -1;
}

int dfs(int v, int t, int pushed) {
    if (pushed == 0) return 0;
    if (v == t) return pushed;
    for (int& cid = ptr[v]; cid < adj[v].size(); ++cid) {
        auto& edge = adj[v][cid];
        int tr = edge.to;
        if (level[v] + 1 != level[tr] || edge.capacity - edge.flow == 0) continue;
        int push = dfs(tr, t, min(pushed, edge.capacity - edge.flow));
        if (push == 0) continue;
        edge.flow += push;
        adj[tr][edge.rev].flow -= push;
        return push;
    }
    return 0;
}

int dinic(int s, int t) {
    int flow = 0;
    while (bfs(s, t)) {
        fill(ptr.begin(), ptr.end(), 0);
        while (int pushed = dfs(s, t, INF)) {
            flow += pushed;
        }
    }
    return flow;
}

int solve_virus_containment(int n, vector<pair<int, int>> connections, vector<int> infected, vector<int> critical) {
    // Check overlap
    set<int> inf_set(infected.begin(), infected.end());
    for (int c : critical) {
        if (inf_set.count(c)) return -1;
    }
    
    int S = n;
    int T = n + 1;
    adj.assign(n + 2, vector<Edge>());
    level.resize(n + 2);
    ptr.resize(n + 2);
    
    // S -> Infected
    for (int i : infected) {
        add_directed_edge(S, i, INF);
    }
    
    // Critical -> T
    for (int c : critical) {
        add_directed_edge(c, T, INF);
    }
    
    // Connections
    for (auto& p : connections) {
        add_edge(p.first, p.second, 1);
    }
    
    return dinic(S, T);
}

int main() {
    // Example 1
    int n = 3;
    vector<pair<int, int>> conn = {{0, 1}, {1, 2}};
    vector<int> inf = {0};
    vector<int> crit = {2};
    
    cout << "Min edges (Ex 1): " << solve_virus_containment(n, conn, inf, crit) << endl;
    
    // Example 2
    int n2 = 5;
    vector<pair<int, int>> conn2 = {{0, 1}, {1, 2}, {0, 3}, {3, 4}, {1, 4}};
    vector<int> inf2 = {0};
    vector<int> crit2 = {2, 4};
    
    cout << "Min edges (Ex 2): " << solve_virus_containment(n2, conn2, inf2, crit2) << endl;
    
    return 0;
}

File: Day51/Part4/virus_containment.py
from collections import deque

def solve_virus_containment(n, connections, infected, critical):
    # Check if any critical node is already infected (should not happen based on constraints)
    infected_set = set(infected)
    for c in critical:
        if c in infected_set:
            return -1
            
    # Build Flow Network
    # Nodes 0 to n-1 are servers.
    # Node n is Source (S).
    # Node n+1 is Sink (T).
    S = n
    T = n + 1
    capacity = [[0] * (n + 2) for _ in range(n + 2)]
    graph = [[] for _ in range(n + 2)]
    
    def add_edge(u, v, cap):
        graph[u].append(v)
        graph[v].append(u) # Residual
        capacity[u][v] = cap
        # capacity[v][u] is 0 initially for directed, but here connections are undirected.
        # Wait, undirected edge (u, v) with capacity 1 means we can flow 1 from u to v OR v to u?
        # Or does it mean we can cut it?
        # If we cut (u, v), we remove the link.
        # In flow network, undirected edge (u, v) is usually modeled as u->v cap 1 AND v->u cap 1.
        # If we saturate u->v, we cut it in that direction.
        # But cutting the physical link stops flow in BOTH directions.
        # However, for separation, if we cut the flow from S to T, we effectively separate them.
        # So yes, u->v cap 1 and v->u cap 1 is correct.
        capacity[v][u] = cap
        
    # Add edges from S to Infected
    for i in infected:
        add_edge(S, i, float('inf'))
        
    # Add edges from Critical to T
    for c in critical:
        add_edge(c, T, float('inf'))
        
    # Add network connections
    for u, v in connections:
        add_edge(u, v, 1)
        
    # Edmonds-Karp Max Flow
    max_flow = 0
    
    while True:
        parent = [-1] * (n + 2)
        queue = deque([S])
        parent[S] = S
        
        path_found = False
        while queue:
            u = queue.popleft()
            if u == T:
                path_found = True
                break
            
            for v in graph[u]:
                if parent[v] == -1 and capacity[u][v] > 0:
                    parent[v] = u
                    queue.append(v)
        
        if not path_found:
            break
            
        # Find bottleneck
        path_flow = float('inf')
        v = T
        while v != S:
            u = parent[v]
            path_flow = min(path_flow, capacity[u][v])
            v = u
            
        max_flow += path_flow
        
        # Update residual capacities
        v = T
        while v != S:
            u = parent[v]
            capacity[u][v] -= path_flow
            capacity[v][u] += path_flow
            v = u
            
    return max_flow

if __name__ == "__main__":
    # Example:
    # 0 -- 1 -- 2
    # 0 is Infected. 2 is Critical.
    # Cut (0, 1) or (1, 2). Min cut 1.
    n = 3
    connections = [(0, 1), (1, 2)]
    infected = [0]
    critical = [2]
    
    print(f"Min edges to remove: {solve_virus_containment(n, connections, infected, critical)}")
    
    # Example 2:
    # 0 -- 1 -- 2
    # |    |
    # 3 -- 4
    # Infected: 0. Critical: 2, 4.
    # Paths: 0-1-2, 0-3-4, 0-1-4? (if 1-4 exists? No)
    # Edges: (0,1), (1,2), (0,3), (3,4), (1,4)
    # Cut (0,1) and (0,3)? Cost 2.
    # Or cut (1,2) and (3,4) and (1,4)? Cost 3.
    # Min cut 2.
    
    n2 = 5
    connections2 = [(0, 1), (1, 2), (0, 3), (3, 4), (1, 4)]
    infected2 = [0]
    critical2 = [2, 4]
    print(f"Min edges to remove (Ex 2): {solve_virus_containment(n2, connections2, infected2, critical2)}")
